WebCryptoAPI/util/helpers.js

```diff
@@ -0,0 +1,240 @@
+    }
+
+    // usages is expected to be provided for a key pair, but we are checking
+    // only a single key. The publicKey and privateKey portions of a key pair
+    // recognize only some of the usages appropriate for a key pair.
+    if (key.type === "public") {
+        ["encrypt", "verify", "wrapKey"].forEach(function(usage) {
+            if (usages.includes(usage)) {
+                correctUsages.push(usage);
+            }
+        });
+    } else if (key.type === "private") {
+        ["decrypt", "sign", "unwrapKey", "deriveKey", "deriveBits"].forEach(function(usage) {
+            if (usages.includes(usage)) {
+                correctUsages.push(usage);
+            }
+        });
+    } else {
+        correctUsages = usages;
+    }
+
+    assert_equals((typeof key.usages), "object", key.type + " key.usages is an object");
+    assert_not_equals(key.usages, null, key.type + " key.usages isn't null");
+
+    // The usages parameter could have repeats, but the usages
+    // property of the result should not.
+    var usageCount = 0;
+    key.usages.forEach(function(usage) {
+        usageCount += 1;
+        assert_in_array(usage, correctUsages, "Has " + usage + " usage");
+    });
+    assert_equals(key.usages.length, usageCount, "usages property is correct");
+}
+
+
+// The algorithm parameter is an object with a name and other
+// properties. Given the name, generate all valid parameters.
+//
+// You can have a "short run" for RSA by including a second, true
+// parameter. Useful if you're testing one variant of the
+// algorithm name exhaustively, so need less exhaustive tests
+// for other variants.
+function allAlgorithmSpecifiersFor(algorithmName, shortRun) {
+    var results = [];
+    var hashes = ["SHA-1", "SHA-256", "SHA-384", "SHA-512"];
+    var keyLengths = [1024, 2048, 3072, 4096];
+    var curves = ["P-256", "P-384", "P-521"];
+
+    if (algorithmName.toUpperCase().substring(0, 3) === "AES") {
+        // Specifier properties are name and length
+        [128, 192, 256].forEach(function(length) {
+            results.push({name: algorithmName, length: length});
+        });
+    } else if (algorithmName.toUpperCase() === "HMAC") {
+        [
+            {name: "SHA-1", length: 160},
+            {name: "SHA-256", length: 256},
+            {name: "SHA-384", length: 384},
+            {name: "SHA-512", length: 512}
+        ].forEach(function(hashAlgorithm) {
+            results.push({name: algorithmName, hash: hashAlgorithm.name, length: hashAlgorithm.length});
+        });
+    } else if (algorithmName.toUpperCase().substring(0, 3) === "RSA") {
+        if (shortRun) {
+            hashes = ["SHA-1", "SHA-256"];
+            keyLengths = [1024];
+        }
+        hashes.forEach(function(hashName) {
+            keyLengths.forEach(function(modulusLength) {
+                [new Uint8Array([3]), new Uint8Array([1,0,1])].forEach(function(publicExponent) {
```

Comment:
There is no need to test any success case here besides 0x010001.  In fact, the case 0x03 should probably be banned because Bleichenbacher.